### ISSUE TITLE
vault mount, plugin exited before we could connect

### ISSUE THREAD

**Issue by Gabology** [hashicorp/vault#3626]
Attempting to mount [my custom plugin](https://github.com/Gabology/vault-gitlab-deploy) in our production environment is failing with the following error:

```
vault mount -path=gitlab -plugin-name=gitlab plugin
Mount error: Error making API request.

URL: POST https://borgias.foo.com/v1/sys/mounts/gitlab
Code: 400. Errors:

* plugin exited before we could connect
```

Testing it running the same Docker image on my local workstation, running the Vault server in dev mode works fine. 

The only two major differences that I can think is that 

1) I have to use `docker cp` to copy the plugin binary into the container as on production the filesystem is mounted with the `noexec` flag. However I tried to do this locally to a container and then it works fine.

2) The host environment:
`Linux borgias 4.4.79+ #1 SMP Wed Nov 8 17:00:14 PST 2017 x86_64 Intel(R) Xeon(R) CPU @ 2.60GHz GenuineIntel GNU/Linux`

**Environment:**

* Vault Version: 0.9.0
* Operating System/Architecture: Google Container Optimized OS

**Vault Config File:**

```
storage "gcs" {
  bucket = "vault-data"
}

listener "tcp" {
  address       = "0.0.0.0:9000"
  tls_cert_file = "/etc/certs/vault.crt"
  tls_key_file  = "/etc/certs/vault.key"
}

plugin_directory = "/vault/file/vault-plugins"
```

**Startup Log Output:**

Posted above

**Expected Behavior:**

It should mount the plugin successfully

**Actual Behavior:**

It fails to mount it. I'm guessing that the plugin might be crashing on startup but I don't seem to get any log output from it about that.

**Steps to Reproduce:**

It's quite hard to reproduce. I'm personally not able to reproduce it locally running the same Docker image on my OSX.

But I guess it should be possible if you start a docker container on Google Container Optimized OS running the `vault:0.9.0` image. Then install my plugin on it.

**Edit:** Running vault with log level debug, this is what I can see from the log output:

```
2017/11/29 12:14:34.118829 [DEBUG] plugin.metadata: starting plugin: path=/vault/file/vault-plugins/vault-gitlab-deploy args=[/vault/file/vault-plugins/vault-gitlab-deploy]
2017/11/29 12:14:34.211875 [DEBUG] plugin.metadata: waiting for RPC address: path=/vault/file/vault-plugins/vault-gitlab-deploy
2017/11/29 12:14:34.216813 [DEBUG] plugin.metadata.vault-gitlab-deploy: 2017/11/29 12:14:34 cannot allocate memory
2017/11/29 12:14:34.216903 [DEBUG] plugin.metadata: plugin process exited: path=/vault/file/vault-plugins/vault-gitlab-deploy
2017/11/29 12:14:34.217011 [ERROR] sys: mount failed: path=gitlab/ error=plugin exited before we could connect
```

**Comment by brianshumate** [comment]
The error `plugin exited before we could connect` definitely implies that the plugin is not able to successfully start. Your mention of filesystems mounted with the `noexec` option certainly sounds like the prime suspect to me in this case.

You can try enabling trace level `-log-level=trace` logging to get more information from Vault around the plugin execution, but this appears to be an environmental issue and not necessarily an issue with Vault itself.

**Comment by Gabology** [comment]
@brianshumate 

Not sure if you saw that but I did run it with `-log-level=trace` this is the output:

```
2017/11/29 12:14:34.118829 [DEBUG] plugin.metadata: starting plugin: path=/vault/file/vault-plugins/vault-gitlab-deploy args=[/vault/file/vault-plugins/vault-gitlab-deploy]
2017/11/29 12:14:34.211875 [DEBUG] plugin.metadata: waiting for RPC address: path=/vault/file/vault-plugins/vault-gitlab-deploy
2017/11/29 12:14:34.216813 [DEBUG] plugin.metadata.vault-gitlab-deploy: 2017/11/29 12:14:34 cannot allocate memory
2017/11/29 12:14:34.216903 [DEBUG] plugin.metadata: plugin process exited: path=/vault/file/vault-plugins/vault-gitlab-deploy
2017/11/29 12:14:34.217011 [ERROR] sys: mount failed: path=gitlab/ error=plugin exited before we could connect
```

**Comment by Gabology** [comment]
I don't think it's a host related issue, if I start a separate Vault container running in dev mode on the same host that actually works.

But nor can I reproduce it locally, I've tried using the same storage provider and the same TLS configuration. 

As soon as I start a container on the same host that is not running Vault in dev mode then it doesn't work.

I've also tried switching distro on the machine to Ubuntu 14 LTS and upgraded the machine from 600MB of RAM to 1.7GB of RAM. But having the same results. Mounting the plugin in a container running vault in dev mode works, but anything else doesn't work.
